orc8r/cloud/go/services/bootstrapper/servicers/registration/storage.go

```diff
@@ -0,0 +1,168 @@
+
+import (
+	"github.com/pkg/errors"
+
+	"magma/orc8r/cloud/go/blobstore"
+	"magma/orc8r/cloud/go/services/bootstrapper"
+	"magma/orc8r/cloud/go/storage"
+	mErrors "magma/orc8r/lib/go/errors"
+	"magma/orc8r/lib/go/protos"
+)
+
+const networkWildcard = "*"
+
+type Store interface {
+	SetTokenInfo(oldNonce string, tokenInfo protos.TokenInfo) error
+	GetTokenInfoFromLogicalID(networkID string, logicalID string) (*protos.TokenInfo, error)
+	GetTokenInfoFromNonce(nonce string) (*protos.TokenInfo, error)
+	IsNonceUnique(nonce string) (bool, error)
+}
+
+type blobstoreStore struct {
+	factory blobstore.StoreFactory
+}
+
+func NewBlobstoreStore(factory blobstore.StoreFactory) Store {
+	return &blobstoreStore{factory}
+}
+
+func (b *blobstoreStore) SetTokenInfo(oldNonce string, tokenInfo protos.TokenInfo) error {
+	networkID := tokenInfo.GatewayPreregisterInfo.NetworkId
+	logicalID := tokenInfo.GatewayPreregisterInfo.LogicalId
+
+	store, err := b.factory.StartTransaction(nil)
+	if err != nil {
+		return err
+	}
+	defer store.Rollback()
+
+	logicalIDBlob, err := tokenInfoToBlob(bootstrapper.LogicalIDToTokenInfo, logicalID, tokenInfo)
+	if err != nil {
+		return err
+	}
+	err = store.Write(networkID, blobstore.Blobs{logicalIDBlob})
+	if err != nil {
+		return err
+	}
+
+	nonceBlob, err := tokenInfoToBlob(bootstrapper.NonceTokenToInfoMap, tokenInfo.Nonce, tokenInfo)
+	if err != nil {
+		return err
+	}
+	err = store.Write(networkWildcard, blobstore.Blobs{nonceBlob})
+	if err != nil {
+		return err
+	}
+
+	if oldNonce != "" {
+		oldNonceTK := storage.TKs{{
+			Type: string(bootstrapper.NonceTokenToInfoMap),
+			Key:  oldNonce,
+		}}
+		err = store.Delete(networkWildcard, oldNonceTK)
+		if err != nil {
+			if err != mErrors.ErrNotFound{
+				return err
+			}
+		}
+	}
+
+	return store.Commit()
+}
+
+func (b *blobstoreStore) GetTokenInfoFromLogicalID(networkID string, logicalID string) (*protos.TokenInfo, error) {
+	store, err := b.factory.StartTransaction(nil)
+	if err != nil {
+		return nil, err
+	}
+	defer store.Rollback()
+
+	logicalIDTK := storage.TK{
+		Type: string(bootstrapper.LogicalIDToTokenInfo),
+		Key:  logicalID,
+	}
+	logicalIDBlob, err := store.Get(networkID, logicalIDTK)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenInfo, err := tokenInfoFromBlob(logicalIDBlob)
+	if err != nil {
+		return nil, err
+	}
+
+	return &tokenInfo, store.Commit()
+}
+
+func (b *blobstoreStore) GetTokenInfoFromNonce(nonce string) (*protos.TokenInfo, error) {
+	store, err := b.factory.StartTransaction(nil)
+	if err != nil {
+		return nil, err
+	}
+	defer store.Rollback()
+
+	nonceTK := storage.TK{
+		Type: string(bootstrapper.NonceTokenToInfoMap),
+		Key:  nonce,
+	}
+	nonceBlob, err := store.Get(networkWildcard, nonceTK)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenInfo, err := tokenInfoFromBlob(nonceBlob)
+	if err != nil {
+		return nil, err
+	}
+
+	return &tokenInfo, store.Commit()
+}
+
+func (b *blobstoreStore) IsNonceUnique(nonce string) (bool, error) {
```

Comment:
The probability of a non-unique nonce is trivial. We do wanna check for it, but it can be done in the "set" method rather than having its own method

Reply:
I had a thought for using this with a couple max retries while generating a nonce in “CloudRegistration”

Reply:
Makes sense. Given the probability + scale we're working with though, it'll be simpler to just throw an error when it does happen, and caller can retry

This is estimated to never happen in practice 🙂